cmake/checks/check_01_cxx_features.cmake

```diff
@@ -51,552 +37,406 @@ ENDIF()
-    # of an undeduced auto return type.
-    #
-    # https://llvm.org/bugs/show_bug.cgi?id=16876
-    #
-    SET(_flags "${DEAL_II_CXX_FLAGS_DEBUG}")
-    STRIP_FLAG(_flags "-Wa,--compress-debug-sections")
-    ADD_FLAGS(CMAKE_REQUIRED_FLAGS "${_flags}")
-    CHECK_CXX_SOURCE_COMPILES(
-      "
-      struct foo
-      {
-        auto func();
-      };
-      int main()
-      {
-        foo bar;
-        (void) bar;
-      }
-      "
-      DEAL_II_HAVE_CXX14_CLANGAUTODEBUG_BUG_OK)
+    int main()
+    {
+      std::condition_variable c;
+      p[0];
+      c.notify_all();
+
+     // type traits functionality
+     constexpr auto m0 = std::is_trivial<double>::value;
+     (void) m0;
+     constexpr auto m1 = std::is_standard_layout<double>::value;
+     (void) m1;
+     constexpr auto m2 = std::is_pod<double>::value;
+     (void) m2;
+    }
+    "
+    DEAL_II_HAVE_CXX11_FEATURES)
 
-    RESET_CMAKE_REQUIRED()
-  ENDIF()
+  # clang libc++ bug, see https://llvm.org/bugs/show_bug.cgi?id=20084
+  CHECK_CXX_SOURCE_COMPILES(
+    "
+    #include <functional>
+    struct A { void foo() const {} };
+    int main() { A a; std::bind(&A::foo,a)(); return 0; }
+    "
+    DEAL_II_HAVE_CXX11_FUNCTIONAL_LLVMBUG20084_OK)
 
-  IF( DEAL_II_HAVE_CXX14_MAKE_UNIQUE AND
-      DEAL_II_HAVE_CXX14_CLANGAUTODEBUG_BUG_OK AND
-      DEAL_II_HAVE_CXX14_CONSTEXPR_STDMAXMIN)
+  IF(DEAL_II_HAVE_CXX14_FEATURES AND
+     DEAL_II_HAVE_CXX14_CLANGAUTODEBUG_BUG_OK AND
+     DEAL_II_HAVE_CXX11_FEATURES AND
+     DEAL_II_HAVE_CXX11_FUNCTIONAL_LLVMBUG20084_OK)
+    MESSAGE(STATUS "C++14 support is enabled.")
     SET(DEAL_II_HAVE_CXX14 TRUE)
   ELSE()
-    IF(NOT _user_provided_cxx_version_flag)
-      SET(DEAL_II_CXX_VERSION_FLAG "")
-    ENDIF()
+    MESSAGE(STATUS "C++14 support is disabled.")
+    SET(DEAL_II_HAVE_CXX14 FALSE)
   ENDIF()
-ENDIF()
+ENDMACRO()
 
 #
-# Check for proper C++11 support.
+# Try to find out what we support:
 #
 
-#
-# Set up a default C++11 flag in case both C++14 detection failed and the user
-# did not specify a flag:
-#
-IF("${DEAL_II_CXX_VERSION_FLAG}" STREQUAL "")
-  _check_version("11" "0x")
-ENDIF()
+_set_up_cmake_required()
 
-UNSET_IF_CHANGED(CHECK_CXX_FEATURES_FLAGS_CXX11_SAVED
-  "${CMAKE_REQUIRED_FLAGS}${DEAL_II_CXX_VERSION_FLAG}"
-  DEAL_II_HAVE_CXX11_FEATURES
-  DEAL_II_HAVE_CXX11_FUNCTIONAL_LLVMBUG20084_OK
-  DEAL_II_HAVE_CXX11_ICCLIBSTDCPP47CXX11BUG_OK
-  DEAL_II_HAVE_CXX11_ICCNUMERICLIMITSBUG_OK
-  DEAL_II_HAVE_CXX11_MACOSXC99BUG_OK
-  )
+_test_cxx14_support()
 
-ADD_FLAGS(CMAKE_REQUIRED_FLAGS "${DEAL_II_CXX_VERSION_FLAG}")
-CHECK_CXX_SOURCE_COMPILES(
-  "
-  // common C++11 include files
-  #include <array>
-  #include <condition_variable>
-  #include <type_traits>
+IF(NOT DEAL_II_HAVE_CXX14)
+  MESSAGE(STATUS "C++14 support not available. Try to set -std=c++14 explicitly")
+  ENABLE_IF_SUPPORTED(DEAL_II_CXX_FLAGS "-std=c++14")
+  ENABLE_IF_SUPPORTED(DEAL_II_CXX_FLAGS "/std:c++14")
```

Comment:
I would just go with `/std:c++latest` here (and see if we still see the problem with non-inline  in class initialized `constexpr` members). If that doesn't work for `Visual Studio 2017`, I would disable `DEAL_II_CXX14_CONSTEXPR_BUG_OK` unconditionally for `Visual Studio 2017`.

Reply:
Yes. The bug is that I don't disable `DEAL_II_CXX14_CONSTEXPR_BUG_OK` unconditionally. I should have look at the logic more carefully.

Reply:
The "/std:c++14" clause actually doesn't get triggered at all I will remove it again.

Reply:
@masterleinad Now I understand. I forgot to fix up the preprocessor macro name yesterday. How silly.